[torch distributed] Implementing all_gather_base (#56304)

liangluofb opened the pull request:
Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/56304

This diff implements the all_gather_base in pytorch distributed.

Test Plan: dist.all_gather_base(output, input)...

Differential Revision: D27488999

zhaojuanmao commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1755,11 +1755,36 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::recvAnysource(
     const AllgatherOptions& /*unused */) {
-  throw std::runtime_error(
-      "no support for allgather_base in NCCL process group");
+  check_gpu_single_tensor(input_tensor);
+  check_gpu_single_tensor(output_tensor);
```
same as above, add a check to make sure output_tensor size is sum of input_tensor size

liangluofb replied:
Same as above,

liangluofb replied:
done. added in cpp

zhaojuanmao reviewed (commented):
@liangluofb thanks for your contribution! the codes overall look good, I left some minor comments. 

Also we've discussed internally, we would like to make 'all_gather_base' to be private APIs and named as '_all_gather_base' in both python and C++. As @agolynski had a C10d API cleanup plans to merge all_gather and all_gather_base. Before that cleanup is done, we would like not to introduce a new set of APIs.  

So would you please help changing the 'all_gather_base' to be  '_all_gather_base' in both python and C++? thanks

liangluofb commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1923,6 +1923,61 @@ def all_gather(tensor_list,
+        tensor([1,2]) # Rank 1
+
+    """
+    _check_single_tensor(input_tensor, "input_tensor")
+    _check_single_tensor(output_tensor, "output_tensor")
```
Hi Yanli, the input tensor size is not known from different ranks. So I am not sure how we test that before allgather_base finishes.

liangluofb replied:
I can add an description that the user is responsible of passing correctly sized output whose length is the sum of all inputs from all ranks.

agolynski commented on torch/lib/c10d/ProcessGroup.hpp:
```diff
@@ -222,14 +222,14 @@ class ProcessGroup : public torch::CustomClassHolder {
   // is interpreted as a contigious collection of size inputBuffer * WORLD_SIZE.
   // For implementers of ProcessGroup API and advanced users only.
-  virtual c10::intrusive_ptr<ProcessGroup::Work> allgather_base(
+  virtual c10::intrusive_ptr<ProcessGroup::Work> _allgather_base(
```
Do we need to rename allgather_base to _allgather_base? E.g. We could just leave allgather_base ASIS, but  name it _all_gather_base in py API to mark it "for private use".

agolynski replied:
discussed offline. I think it's fine to rename it here, let's add the corresponding comment so users have a better idea re: future of this function?

agolynski commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1754,12 +1754,48 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::recvAnysource(
+  check_gpu_single_tensor(output_tensor);
+
+  if(input_tensor.dtype() != output_tensor.dtype())
+  {
```
nit: let's use consistent formatting, e.g.
if(input_tensor.dtype() != output_tensor.dtype()) {

here and below

agolynski commented on torch/lib/c10d/ProcessGroup.hpp:
```diff
@@ -222,14 +222,14 @@ class ProcessGroup : public torch::CustomClassHolder {
   // Gathers a single tensor inputBuffer into a single buffer outputBuffer that
   // is interpreted as a contigious collection of size inputBuffer * WORLD_SIZE.
   // For implementers of ProcessGroup API and advanced users only.
```
Could you update the comment saying this function will be deprecated going forward?

agolynski reviewed (commented):
Mostly LG, just some minor comments thank you!

agolynski commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1783,12 +1783,46 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::recvAnysource(
+  check_gpu_single_tensor(input_tensor);
+  check_gpu_single_tensor(output_tensor);
+
+  if(input_tensor.dtype() != output_tensor.dtype()) {
```
formatting nit: "if ("

liangluofb committed:
[torch distributed] Implementing all_gather_base (#56315)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/56315

This diff implements the all_gather_base in pytorch distributed.

Test Plan: dist.all_gather_base(output, input)...

Reviewed By: agolynski, amylittleyang

Differential Revision: D27488999

fbshipit-source-id: 4b286563301324f9212c0e31f8540712453441e2

zarzen commented:
Hi, 
would you mind to change the function signature of _allgather_base to accept two list tensors? so that multiple tensor allgather can be launched in a single ncclGroup? and the current `collective` interface requires `std::vector<at::Tensor>` as inputs anyway. 
I can submit a PR if you would agree on the change. 
```cpp
c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::_allgather_base(
     std::vector<at::Tensor>& outputs,
    std::vector<at::Tensor>& inputs,
    const AllgatherOptions& /*unused */)

```
At the python front end we can do a type checking, if the input is a single tensor then wrap it with the brackets.
```python
def allgather_base(self, outputs, inputs, ... ):
    if type(output) != list:
        outputs = [outputs]
        inputs = [inputs]
    ...
```
